applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp

```diff
@@ -8,20 +8,166 @@
+//         KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ACCELERATION,rnode)
+
+        KRATOS_CHECK_DOF_IN_NODE(DISPLACEMENT_X, rnode)
+        KRATOS_CHECK_DOF_IN_NODE(DISPLACEMENT_Y, rnode)
+        KRATOS_CHECK_DOF_IN_NODE(DISPLACEMENT_Z, rnode)
+    }
+
+    // Verify that the constitutive law exists
+    KRATOS_ERROR_IF_NOT(pElement->GetProperties().Has( CONSTITUTIVE_LAW )) << "Constitutive law not provided for property " << pElement->GetProperties().Id() << std::endl;
+
+    // Verify that the constitutive law has the correct dimension
+    const SizeType strain_size = pElement->GetProperties().GetValue( CONSTITUTIVE_LAW )->GetStrainSize();
+    if ( dimension == 2 ) {
+        KRATOS_ERROR_IF( strain_size < 3 || strain_size > 4) << "Wrong constitutive law used. This is a 2D element! expected strain size is 3 or 4 (el id = ) " << pElement->Id() << std::endl;
+    } else {
+        KRATOS_ERROR_IF_NOT(strain_size == 6) << "Wrong constitutive law used. This is a 3D element! expected strain size is 6 (el id = ) "<<  pElement->Id() << std::endl;
+    }
+
+    return 0;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ComputeEquivalentF(
+    const Element* pElement,
+    Matrix& rF,
+    const Vector& rStrainTensor
+    )
+{
+    const auto& r_geometry = pElement->GetGeometry();
+    const SizeType dimension = r_geometry.WorkingSpaceDimension();
+
+    if(dimension == 2) {
+        rF(0,0) = 1.0+rStrainTensor(0);
+        rF(0,1) = 0.5*rStrainTensor(2);
+        rF(1,0) = 0.5*rStrainTensor(2);
+        rF(1,1) = 1.0+rStrainTensor(1);
+    } else {
+        rF(0,0) = 1.0+rStrainTensor(0);
+        rF(0,1) = 0.5*rStrainTensor(3);
+        rF(0,2) = 0.5*rStrainTensor(5);
+        rF(1,0) = 0.5*rStrainTensor(3);
+        rF(1,1) = 1.0+rStrainTensor(1);
+        rF(1,2) = 0.5*rStrainTensor(4);
+        rF(2,0) = 0.5*rStrainTensor(5);
+        rF(2,1) = 0.5*rStrainTensor(4);
+        rF(2,2) = 1.0+rStrainTensor(2);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void CalculateB(
+    const Element* pElement,
+    Matrix& rB,
+    const Matrix& rDN_DX
+    )
+{
+    const auto& r_geometry = pElement->GetGeometry();
+    const SizeType number_of_nodes = r_geometry.PointsNumber();
+    const SizeType dimension = r_geometry.WorkingSpaceDimension();
+
+    rB.clear();
+
+    if(dimension == 2) {
+        for ( SizeType i = 0; i < number_of_nodes; ++i ) {
+            rB(0, i*2    ) = rDN_DX(i, 0);
+            rB(1, i*2 + 1) = rDN_DX(i, 1);
+            rB(2, i*2    ) = rDN_DX(i, 1);
+            rB(2, i*2 + 1) = rDN_DX(i, 0);
+        }
+    } else if(dimension == 3) {
+        for ( SizeType i = 0; i < number_of_nodes; ++i ) {
+            rB(0, i*3    ) = rDN_DX(i, 0);
```

Comment:
obvious optimization: compute i*3 at the beginning of the loop

Reply:
In the future PR